### PR TITLE
Upgrade coin to v3 for better events

### PR DESCRIPTION
**Pull request by sirlensalot**

### CODE REVIEW

File: test/Chainweb/Test/Pact/PactInProcApi.hs [hunk 1]
```diff
@@ -256,6 +272,138 @@ newBlockRewindValidate mpRefIO reqIO = testCase "newBlockRewindValidate" $ do
...
+
+  tx21_2 <- txResult 2 pwo21
+  gasEv2 <- mkTransferEvent "sender00" "NoMiner" 0.0014 "coin" v3Hash
+  sendTfr <- mkTransferEvent "sender00" "" 0.0123 "coin" v3Hash
+  assertEqual "Events for tx2 @ block 21" [gasEv2,sendTfr] (_crEvents tx21_2)
+
+
+
+
+
```
Comment (larskuhtz) [inline test/Chainweb/Test/Pact/PactInProcApi.hs:332]:
```suggestion




```

**Review by larskuhtz — APPROVED** [review]
I looked only at the Haskell code. I didn't review the pact code.

File: pact/coin-contract/v3/coin-v3.pact [hunk 2]
```diff
@@ -0,0 +1,619 @@
...
+    [ (defproperty conserves-mass
+        (= (column-delta coin-table 'balance) 0.0))
+
+      (defproperty valid-account (account:string)
+        (and
+          (>= (length account) 3)
+          (<= (length account) 256)))
+    ]
+
+  (implements fungible-v2)
```
[the remark this answers is not in the document]
Comment (sirlensalot) [reply pact/coin-contract/v3/coin-v3.pact:20]: Yeah, no changes to `fungible-v2`

File: pact/coin-contract/v3/coin-v3.pact [hunk 3]
```diff
@@ -0,0 +1,619 @@
...
+
+  (bless "ut_J_ZNkoyaPUEJhiwVeWnkSQn9JT9sQCWKdjjVVrWo")
```
[the remark this answers is not in the document]
Comment (sirlensalot) [reply pact/coin-contract/v3/coin-v3.pact:22]: yes. I'm not sure about extensive commenting in onchain contracts, what do you think?

[the remark this answers is not in the document]
Comment (emilypi) [reply pact/coin-contract/v3/coin-v3.pact:22]: Probably fine to leave as is

File: pact/coin-contract/v3/coin-v3.pact [hunk 4]
```diff
@@ -0,0 +1,630 @@
...
+
+    (step (buy-gas sender total))
+    (step (redeem-gas miner miner-guard sender total))
+    )
+
+  (defun debit:string (account:string amount:decimal)
+    @doc "Debit AMOUNT from ACCOUNT balance"
+
+    @model [ (property (> amount 0.0))
+             (property (valid-account account))
```
Comment (emmanueldenloye) [inline pact/coin-contract/v3/coin-v3.pact:406]: Why do you do call `validate-account` in the property test and in the function body?

**Review by emmanueldenloye — APPROVED** [review]
I simply just have those questions detailed in the accompanying comments. Otherwise, LGTM.

File: test/Chainweb/Test/Pact/PactInProcApi.hs [hunk 5]
```diff
@@ -91,12 +103,20 @@ tests rdb = ScheduledTest testName $ go
...
          , test Warn $ moduleNameFork
+         , multiChainTest "pact4coin3UpgradeTest" pact4coin3UpgradeTest
          ]
       where
         test logLevel f =
           withDelegateMempool $ \dm ->
           withPactTestBlockDb testVersion cid logLevel rdb (snd <$> dm) defaultPactServiceConfig $
           f (fst <$> dm)
+
+        multiChainTest tname f =
```
[the remark this answers is not in the document]
Comment (sirlensalot) [reply test/Chainweb/Test/Pact/PactInProcApi.hs:114]: Let's see if there's more tests like this before worrying about logging func.

**Review by emmanueldenloye — APPROVED** [review]
LGTM

**Pull request merged**